src/ray/gcs/redis_module/ray_redis_module.cc

```diff
@@ -518,6 +532,142 @@ int SetRemove_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int ar
+///
+/// \param notification_mode Output the mode of the operation: APPEND_OR_ADD or REMOVE.
+/// \param deleted_data Output data if the deleted data is not the same as required.
+int HashUpdate_DoWrite(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
+                       GcsTableNotificationMode *notification_mode,
+                       RedisModuleString *&deleted_data) {
+  if (argc != 5) {
+    return RedisModule_WrongArity(ctx);
+  }
+  RedisModuleString *prefix_str = argv[1];
+  RedisModuleString *id = argv[3];
+  RedisModuleString *update_data = argv[4];
+
+  RedisModuleKey *key;
+  REPLY_AND_RETURN_IF_NOT_OK(OpenPrefixedKey(
+      &key, ctx, prefix_str, id, REDISMODULE_READ | REDISMODULE_WRITE, nullptr));
+  int type = RedisModule_KeyType(key);
+  REPLY_AND_RETURN_IF_FALSE(
+      type == REDISMODULE_KEYTYPE_HASH || type == REDISMODULE_KEYTYPE_EMPTY,
+      "HashUpdate_DoWrite: entries must be a hash or an empty hash");
+
+  size_t update_data_len = 0;
+  const char *update_data_buf = RedisModule_StringPtrLen(update_data, &update_data_len);
+
+  auto data_vec = flatbuffers::GetRoot<GcsTableEntry>(update_data_buf);
+  *notification_mode = data_vec->notification_mode();
+  if (*notification_mode == GcsTableNotificationMode::APPEND_OR_ADD) {
+    // This code path means they are updating command.
+    size_t total_size = data_vec->entries()->size();
+    REPLY_AND_RETURN_IF_FALSE(total_size % 2 == 0, "Invalid Hash Update data vector.");
+    for (int i = 0; i < total_size; i += 2) {
+      // Reconstruct a key-value pair from a flattened list.
+      RedisModuleString *entry_key = RedisModule_CreateString(
+          ctx, data_vec->entries()->Get(i)->data(), data_vec->entries()->Get(i)->size());
+      RedisModuleString *entry_value =
+          RedisModule_CreateString(ctx, data_vec->entries()->Get(i + 1)->data(),
+                                   data_vec->entries()->Get(i + 1)->size());
+      // Returning 0 if key exists(still updated), 1 if the key is created.
+      RAY_IGNORE_EXPR(
+          RedisModule_HashSet(key, REDISMODULE_HASH_NONE, entry_key, entry_value, NULL));
+    }
+  } else {
+    // This code path means the command wants to remove the entries.
+    // Hash_DoPublish only publish the entries that have been removed.
+    size_t total_size = data_vec->entries()->size();
+    std::vector<bool> deleted_flags(total_size, true);
+    int remove_count = 0;
+    for (int i = 0; i < total_size; i++) {
+      RedisModuleString *entry_key = RedisModule_CreateString(
+          ctx, data_vec->entries()->Get(i)->data(), data_vec->entries()->Get(i)->size());
+      int deleted_num = RedisModule_HashSet(key, REDISMODULE_HASH_NONE, entry_key,
+                                            REDISMODULE_HASH_DELETE, NULL);
+      deleted_flags[i] = deleted_num != 0;
+      remove_count += deleted_num;
+    }
+    if (remove_count != deleted_flags.size()) {
```

Comment:
Maybe it's a little slower, but I would consider removing this if-statement and always creating the `GcsTableEntry` here no matter how many keys were successfully deleted, for code clarity. It may also make sense to do this even for additions, when the operation should always succeed.

Reply:
Done.

Reply:
Hmm it seems like this is still the original code?